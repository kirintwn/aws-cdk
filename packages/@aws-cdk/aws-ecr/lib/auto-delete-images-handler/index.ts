// eslint-disable-next-line import/no-extraneous-dependencies
import { ECR } from 'aws-sdk';

const ecr = new ECR();

export async function handler(event: AWSLambda.CloudFormationCustomResourceEvent) {
  switch (event.RequestType) {
    case 'Create':
    case 'Update':
      return;
    case 'Delete':
      return onDelete(event);
  }
}

/**
 * Recursively delete all images in the repository
 *
 * @param ECR.ListImagesRequest the repositoryName & nextToken if presented
 */
async function emptyRepository(params: ECR.ListImagesRequest) {
  const listedImages = await ecr.listImages(params).promise();
  const imageIds = listedImages?.imageIds ?? [];
  const nextToken = listedImages.nextToken ?? null;
  if (imageIds.length === 0) {
    return;
  }

  await ecr.batchDeleteImage({
    repositoryName: params.repositoryName,
    imageIds,
  }).promise();

  if (nextToken) {
    await emptyRepository({
      ...params,
      nextToken,
    });
  }
}

async function onDelete(deleteEvent: AWSLambda.CloudFormationCustomResourceDeleteEvent) {
  const repositoryName = deleteEvent.ResourceProperties?.RepositoryName;
  if (!repositoryName) {
    throw new Error('No RepositoryName was provided.');
  }
  await emptyRepository({ repositoryName });
}
