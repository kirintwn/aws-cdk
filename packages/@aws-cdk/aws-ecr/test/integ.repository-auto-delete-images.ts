/// !cdk-integ pragma:ignore-assets
import * as cdk from '@aws-cdk/core';
import * as ecr from '../lib';

const app = new cdk.App();
const stack = new cdk.Stack(app, 'aws-ecr-integ-stack');

const repo = new ecr.Repository(stack, 'Repo', {
  repositoryName: 'delete-even-if-containing-images',
  removalPolicy: cdk.RemovalPolicy.DESTROY,
  autoDeleteImages: true,
});

new cdk.CfnOutput(stack, 'RepositoryURI', {
  value: repo.repositoryUri,
});

app.synth();
