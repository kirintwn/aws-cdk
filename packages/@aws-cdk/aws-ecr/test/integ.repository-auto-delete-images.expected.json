{
  "Resources": {
    "Repo02AC86CF": {
      "Type": "AWS::ECR::Repository",
      "Properties": {
        "RepositoryName": "delete-even-if-containing-images",
        "RepositoryPolicyText": {
          "Statement": [
            {
              "Action": [
                "ecr:BatchDeleteImage",
                "ecr:ListImages"
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::GetAtt": [
                    "CustomECRAutoDeleteImagesCustomResourceProviderRole665F2773",
                    "Arn"
                  ]
                }
              }
            }
          ],
          "Version": "2012-10-17"
        }
      },
      "UpdateReplacePolicy": "Delete",
      "DeletionPolicy": "Delete"
    },
    "RepoAutoDeleteImagesCustomResource65201E29": {
      "Type": "Custom::ECRAutoDeleteImages",
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomECRAutoDeleteImagesCustomResourceProviderHandler8D89C030",
            "Arn"
          ]
        },
        "RepositoryName": {
          "Ref": "Repo02AC86CF"
        }
      },
      "DependsOn": [
        "Repo02AC86CF"
      ],
      "UpdateReplacePolicy": "Delete",
      "DeletionPolicy": "Delete"
    },
    "CustomECRAutoDeleteImagesCustomResourceProviderRole665F2773": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Version": "2012-10-17",
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ]
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
          }
        ]
      }
    },
    "CustomECRAutoDeleteImagesCustomResourceProviderHandler8D89C030": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "AssetParameters47ae0d4fa818731652e6346a46cc94926230fc830387c287b0668ff8a52cdf1cS3Bucket3292363C"
          },
          "S3Key": {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::Select": [
                    0,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters47ae0d4fa818731652e6346a46cc94926230fc830387c287b0668ff8a52cdf1cS3VersionKey9ADF5496"
                        }
                      ]
                    }
                  ]
                },
                {
                  "Fn::Select": [
                    1,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters47ae0d4fa818731652e6346a46cc94926230fc830387c287b0668ff8a52cdf1cS3VersionKey9ADF5496"
                        }
                      ]
                    }
                  ]
                }
              ]
            ]
          }
        },
        "Timeout": 900,
        "MemorySize": 128,
        "Handler": "__entrypoint__.handler",
        "Role": {
          "Fn::GetAtt": [
            "CustomECRAutoDeleteImagesCustomResourceProviderRole665F2773",
            "Arn"
          ]
        },
        "Runtime": "nodejs12.x"
      },
      "DependsOn": [
        "CustomECRAutoDeleteImagesCustomResourceProviderRole665F2773"
      ]
    }
  },
  "Parameters": {
    "AssetParameters47ae0d4fa818731652e6346a46cc94926230fc830387c287b0668ff8a52cdf1cS3Bucket3292363C": {
      "Type": "String",
      "Description": "S3 bucket for asset \"47ae0d4fa818731652e6346a46cc94926230fc830387c287b0668ff8a52cdf1c\""
    },
    "AssetParameters47ae0d4fa818731652e6346a46cc94926230fc830387c287b0668ff8a52cdf1cS3VersionKey9ADF5496": {
      "Type": "String",
      "Description": "S3 key for asset version \"47ae0d4fa818731652e6346a46cc94926230fc830387c287b0668ff8a52cdf1c\""
    },
    "AssetParameters47ae0d4fa818731652e6346a46cc94926230fc830387c287b0668ff8a52cdf1cArtifactHash59CA20D1": {
      "Type": "String",
      "Description": "Artifact hash for asset \"47ae0d4fa818731652e6346a46cc94926230fc830387c287b0668ff8a52cdf1c\""
    }
  },
  "Outputs": {
    "RepositoryURI": {
      "Value": {
        "Fn::Join": [
          "",
          [
            {
              "Fn::Select": [
                4,
                {
                  "Fn::Split": [
                    ":",
                    {
                      "Fn::GetAtt": [
                        "Repo02AC86CF",
                        "Arn"
                      ]
                    }
                  ]
                }
              ]
            },
            ".dkr.ecr.",
            {
              "Fn::Select": [
                3,
                {
                  "Fn::Split": [
                    ":",
                    {
                      "Fn::GetAtt": [
                        "Repo02AC86CF",
                        "Arn"
                      ]
                    }
                  ]
                }
              ]
            },
            ".",
            {
              "Ref": "AWS::URLSuffix"
            },
            "/",
            {
              "Ref": "Repo02AC86CF"
            }
          ]
        ]
      }
    }
  }
}